#!/usr/bin/env nextflow
nextflow.enable.dsl = 2

/*
 * WGS Variant Calling Workflow
 * Generated by BioPipelines Workflow Composer
 * Date: 2025-11-25
 */

// Import modules
include { FASTQC } from './nextflow-modules/fastqc/main.nf'
include { BWAMEM_ALIGN } from './nextflow-modules/bwamem/main.nf'
include { GATK_HAPLOTYPECALLER } from './nextflow-modules/gatk_haplotypecaller/main.nf'
include { PICARD_MARKDUPLICATES } from './nextflow-modules/markduplicates/main.nf'

// Parameters
params.reads = "data/raw/*_R{1,2}.fastq.gz"
params.genome = "data/references/GRCh38/genome.fa"
params.known_sites = "data/references/dbsnp.vcf.gz"
params.outdir = "results"

// Main workflow
workflow {
    // Input channels
    reads_ch = Channel
        .fromFilePairs(params.reads, checkIfExists: true)
        .map { sample_id, reads -> tuple([id: sample_id], reads) }
    
    genome = file(params.genome, checkIfExists: true)
    known_sites = file(params.known_sites, checkIfExists: true)
    
    // QC
    FASTQC(reads_ch)
    
    // Build index
    BWA_INDEX(genome)
    
    // Align
    BWA_MEM(reads_ch, BWA_INDEX.out.index)
    
    // Sort
    SAMTOOLS_SORT(BWA_MEM.out.bam)
    
    // Mark duplicates
    GATK_MARKDUPLICATES(SAMTOOLS_SORT.out.bam)
    
    // Base recalibration
    GATK_BASERECALIBRATOR(GATK_MARKDUPLICATES.out.bam, genome, known_sites)
    GATK_APPLYBQSR(GATK_MARKDUPLICATES.out.bam, GATK_BASERECALIBRATOR.out.table, genome)
    
    // Variant calling
    GATK_HAPLOTYPECALLER(GATK_APPLYBQSR.out.bam, genome, Channel.empty())
    
    // Filter variants
    BCFTOOLS_FILTER(GATK_HAPLOTYPECALLER.out.vcf, "QUAL>30 && DP>10")
    
    // Stats
    BCFTOOLS_STATS(BCFTOOLS_FILTER.out.vcf)
    
    // QC aggregation
    MULTIQC(FASTQC.out.zip.mix(BCFTOOLS_STATS.out.stats).collect())
}
