#!/usr/bin/env nextflow
nextflow.enable.dsl = 2

/*
 * RNA-seq Differential Expression Workflow
 * Generated by BioPipelines Workflow Composer
 * Date: 2025-11-25
 */

// Import modules
include { FASTQC } from './nextflow-modules/fastqc/main.nf'
include { STAR_ALIGN } from './nextflow-modules/star/main.nf'
include { FEATURECOUNTS } from './nextflow-modules/featurecounts/main.nf'

// Parameters
params.reads = "data/raw/*_R{1,2}.fastq.gz"
params.genome = "data/references/GRCm39/genome.fa"
params.gtf = "data/references/GRCm39/genes.gtf"
params.outdir = "results"
params.samplesheet = "samplesheet.csv"

// Main workflow
workflow {
    // Input channels
    reads_ch = Channel
        .fromFilePairs(params.reads, checkIfExists: true)
        .map { sample_id, reads -> tuple([id: sample_id], reads) }
    
    genome = file(params.genome, checkIfExists: true)
    gtf = file(params.gtf, checkIfExists: true)
    
    // QC raw reads
    FASTQC(reads_ch)
    
    // Build STAR index
    STAR_INDEX(genome, gtf)
    
    // Align reads
    STAR_ALIGN(reads_ch, STAR_INDEX.out.index, gtf)
    
    // Count features
    FEATURECOUNTS(STAR_ALIGN.out.bam, gtf)
    
    // Collect counts for DE analysis
    counts_ch = FEATURECOUNTS.out.counts.collect()
    
    // Differential expression
    DESEQ2_DIFFERENTIAL(
        counts_ch,
        file(params.samplesheet),
        "condition",
        "treatment",
        "control"
    )
    
    // Aggregate QC
    MULTIQC(
        FASTQC.out.zip
            .mix(STAR_ALIGN.out.log)
            .mix(FEATURECOUNTS.out.summary)
            .collect()
    )
}
