#!/usr/bin/env nextflow
nextflow.enable.dsl = 2

/*
 * ChIP-seq Peak Calling Workflow
 * Generated by BioPipelines Workflow Composer
 * Date: 2025-11-25
 */

// Import modules
include { FASTQC } from './nextflow-modules/fastqc/main.nf'
include { MACS2_CALLPEAK } from './nextflow-modules/macs2/main.nf'

// Parameters
params.treatment = "data/chip/*_treatment_R{1,2}.fastq.gz"
params.control = "data/chip/*_input_R{1,2}.fastq.gz"
params.genome = "data/references/GRCh38/genome.fa"
params.outdir = "results"

// Main workflow
workflow {
    // Input channels
    treatment_ch = Channel
        .fromFilePairs(params.treatment, checkIfExists: true)
        .map { sample_id, reads -> tuple([id: sample_id, type: 'treatment'], reads) }
    
    control_ch = Channel
        .fromFilePairs(params.control, checkIfExists: true)
        .map { sample_id, reads -> tuple([id: sample_id, type: 'control'], reads) }
    
    genome = file(params.genome, checkIfExists: true)
    
    // QC
    FASTQC(treatment_ch.mix(control_ch))
    
    // Build index
    BWA_INDEX(genome)
    
    // Align reads
    BWA_MEM(treatment_ch.mix(control_ch), BWA_INDEX.out.index)
    
    // Sort and index BAMs
    SAMTOOLS_SORT(BWA_MEM.out.bam)
    SAMTOOLS_INDEX(SAMTOOLS_SORT.out.bam)
    
    // Separate treatment and control
    treatment_bam = SAMTOOLS_SORT.out.bam
        .filter { meta, bam -> meta.type == 'treatment' }
    control_bam = SAMTOOLS_SORT.out.bam
        .filter { meta, bam -> meta.type == 'control' }
    
    // Peak calling
    MACS2_CALLPEAK(treatment_bam, control_bam, 'narrow', 'hs')
    
    // Motif analysis
    HOMER_FINDMOTIFSGENOME(MACS2_CALLPEAK.out.peaks, genome, 200)
    
    // QC aggregation
    MULTIQC(
        FASTQC.out.zip
            .mix(MACS2_CALLPEAK.out.qc)
            .collect()
    )
}
